url063b9533f6_karlney_tetris_ai_tgz-pJ8-karlney_tetris_cmd_TetrisGameRunnerJ8/src/karlney/tetris/core/Player.java

```diff
@@ -0,0 +1,185 @@
+        t.interrupt();
+    }
+
+    public synchronized void setLevel(int level) {
+        this.level = level;
+    }
+
+
+    public synchronized void updatePlaceScore(int removedRows){
+        score+= ((21+(3*level))-freeFallIterations+1);
+
+        if	(removedRows==1)
+            score=score+(level)*50;
+        if	(removedRows==2)
+            score=score+(level)*150;
+        if	(removedRows==3)
+            score=score+(level)*300;
+        if	(removedRows==4)
+            score=score+(level)*1000;
+
+        lines +=removedRows;
+
+        if (lines>0 && removedRows>0 && (lines)%5000==0){
+            log.info("Lines cleared "+lines+". Pieces played "+getNumberOfPieces()+" Pieces/Lines "+(double)getNumberOfPieces()/(lines));
+        }
+    }
+
+    protected synchronized void commitCurrentPieceToBoard() throws UnableToPlacePieceException {
+        board.placePieceOnBoard(currentPiece);
+        int rows= board.removeFullRows();
+        updatePlaceScore(rows);
+        currentPiece = nextPiece;
+        nextPiece    = generator.getNextPiece(board);
+        freeFallIterations = 0;
+        updateDistribution(currentPiece.getSquare(0, 0).getType());
+        //Make sure we have not reached end of game
+        if (!board.allowedPlacement(currentPiece)){
+            throw new UnableToPlacePieceException("The next piece "+currentPiece+" could not be placed on the board. This means GAME OVER.",null);
+        }
+    }
+
+    private synchronized void updateDistribution(PieceType type) {
+        if (pieceStatistics.containsKey(type)){
+            pieceStatistics.put(type,pieceStatistics.get(type)+1);
+        }
+    }
+
+    public void processInput(PlayerInput input){
+        switch (input){
+            case ROTATE: currentPiece.rotateIfPossible(); break;
+            case DOWN: t.interrupt(); break;
+            case LEFT: currentPiece.moveSideWays(input); break;
+            case RIGHT: currentPiece.moveSideWays(input); break;
+            case DROP: currentPiece.drop(); t.interrupt(); break;
+        }
+    }
+
+    public boolean isRunning() {
+        return running;
+    }
+
+    public void run(){
+        while (running){
+            try {
+                if (currentPiece.isDropped()){
+                    Thread.sleep(FALLING_DELAY);
+                }else{
+                    freeFallIterations++;
+                    Thread.sleep(delay);
+                }
+            }catch(InterruptedException e) {}
+            moveDownOrLand();
+        }
+    }
+
+
+    public Piece getNextPiece() {
+        return nextPiece;
+    }
+
+    public Piece getCurrentPiece() {
+        return currentPiece;
+    }
+
+    public Board getBoard() {
+        return board;
+    }
+
+    public long getScore() {
+        return score;
+    }
+
+    public int getLines() {
+        return lines;
+    }
+
+    public int getNumberOfPieces() {
+        int out = 0;
+        for (Integer i: pieceStatistics.values()){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 This line may not be thread-safe: **179**. At this line the following field is not protected by synchronization: `pieceStatistics`. Your code typically protects this field with synchronization on object: `Player.this`.

**Fix**  
 Consider protecting the above field with synchronization on object: `Player.this`. For example, like at lines: **124**, **123**, and **124**.

**More info**  
 The following lines are where the above field is already protected by synchronization on the above object: **123**, **124**, and **124**.